storybookjs/frontpage
# Add docs versioning

>**Note for reviewers**
> 
> This is a much-improved take on #298. This new approach requires _only_ updating releases files and Netlify branch build settings; everything else is automatic (based on the monorepo). 😎 
> 
> I've taken care to organize the commits such that reviewing commit-by-commit should be clear (and doing so is recommended).

### Description

**Add versioning to Storybook docs.**

- Add a workflow to respond to monorepo events dispatched and create version branches
    - Responds to event dispatched by monorepo workflow
    - See: storybookjs/storybook#16580
- Via Netlify's branch deploys, build an instance of the docs portion of the site referencing specific versions from the monorepo.
- Use Netlify's redirects to proxy the versioned URL to the branch build and Netlify's headers to prevent non-latest pages from being indexed by search engines.
- Update URLs throughout to include version (unless latest)
    - Ensure relative links, canonical URLs, and social graph images are not broken
- Add a version selector control to the docs sidebar to allow switching between these versions.
- When not viewing latest docs, add a banner pointing to the latest version.
- Other minor changes
    - Remove version from Header
    - Remove "v" prefix from `latestVersion`
    - Only build addons pages when referencing latest version
    - Add "Publishing new versions" section to README

### Screenshot

![Docs page showing the version in the url, version selector in the sidebar, and version callout above the content](https://user-images.githubusercontent.com/486540/139288859-a51998ac-6d40-4a2a-ba29-29e563bd1877.png)

### How to Test

1. Check Chromatic tests
1. Run the site locally (`yarn start` & `yarn build`)
    - Ensure the correct version of docs is built and you can navigate fully
        - Keep in mind the redirects won't work
1. Run the site locally with a reference branch specified
    - `BRANCH=release-6-0 yarn build`, or `BRANCH=release-6-0 yarn extract-monorepo-docs && yarn start`
    - Ensure the correct version of docs is built and you can navigate fully
        - Keep in mind the redirects won't work
1. Ensure that `yarn extract-monorepo-docs` and `yarn link-monorepo-docs` still work, with their various arguments
1. Triple-check anything SEO-related (redirect generation code, canonical URLs, headers), please 😉 

#### TODO

- [ ] Ensure algolia search works w/ and w/o versions
    - Fairly sure it works w/o
    - They provided a path for versioning support, but will wait until we get on their latest beta release

After merging this...
- [ ] Add `release-6-0`, `release-6-1`, `release-6-2`, `release-6-4` branches to Netlify build settings

After merging https://github.com/storybookjs/storybook/pull/16580...
- [ ] Push to those branches in the monorepo to kick off the process
